### DISCUSSION TITLE
How to make this project more user user-friendly? Which features would you like to be implemented first?

### DISCUSSION THREAD

**Discussion by filips123** [filips123/PWAsForFirefox#116] [General]
## Call for Discussion

**TL;DR:** How to make this project more user-friendly? What would you like to be improved? Any other suggestions?

PWAsForFirefox is now around 6 months old, and in this time, it has gained some users and attraction. However, there are still some problems/bugs that cause some people not to use it, and it appears some people find the documentation and usage instructions a bit confusing. So, I'm opening this discussion in the hope to resolve some of those problems and make it more user-friendly.

Please use this discussion to suggest any thoughts or ideas how to improve this project. Feel free to also create a new issue or discussion if you have a bigger suggestion, but otherwise, you can just leave a comment here, and I'll create a new issue if needed. Of course, I will also appreciate it if you could help by contributing PRs for those features and fixes.

To start the discussion, I'll give you some of my thoughts about the project and its problems, and possible improvements to address this.

### Better documentation & Website

**Update:** A new documentation website is in progress. Please check #335 for details.

Although this repository already has [a wiki](https://github.com/filips123/PWAsForFirefox/wiki), it seems it is still not enough to make PWAsForFirefox easier for less experienced users. I'm not sure if this is because some users just haven't found it, or is it because it is still too technical for them. So, to address the problems with confusing documentation, I've been thinking about creating some "fancy" website that would better explain what this project is and how to use it, and hopefully make it more appealing and easier for less experienced users (in this case, the wiki will probably move its content to the website).

I think this website should be static and probably hosted on GitHub Pages, but I'm not sure yet whether to use GitHub-provided Jekyll or an alternative static site generator, or instead use Read the Docs. Read the Docs has the advantage of versioned documentation, which might be useful, but using classic site generators could make the website feel more user-friendly.

What do you think? Will the website solve some documentation problems? How should that website be structured? Which parts of the documentation would you like to be improved and/or hosted on the website?

### Easier Setup

After observing some existing discussions about the project, it seems some users feel this project is hard and confusing to set up. I understand that, because unlike browsers with built-in PWA support, you have to install the extension and an additional program to make this project work. Unfortunately, I already made the setup process as easy as I could (it is basically just one setup wizard where almost everything is automated), so there probably won't be any big improvements coming (except if someone has a good suggestion how to improve it).

There are still a few things related to the setup process that could be improved:

* **Easier changing of installation paths (#109) & Installing the project without admin privileges (#111):**

  The first issue is specifically about changing the path where Firefox runtime is installed using environment variables (to enable making it global for all users). Because Firefox runtime is downloaded and copied to the correct directory automatically, this shouldn't be too hard to implement, so I will probably implement this in the near future.

  However, changing other paths (for example, the location of UserChrome source, executables, profiles and config, which could be used to install the project without admin privileges) is harder, because those files need to be copied to the correct directory before the program can function (which is normally done by MSI installer and packages for other operating systems). It is possible to use [build-time environment variables](https://github.com/filips123/PWAsForFirefox/blob/main/native/src/directories.rs#L23-L70) to configure those paths, but this isn't very user-friendly, as it requires building the project from source and manually copying files to the correct location. Because I'm aware of any way to do this easily and automatically, this problem currently has a low priority for me, but I would still like to hear some ideas how to solve it.

* **Less intrusive updates (#114):**

  Currently, every time the extension or the native connector gets updated, the browser will open a new tab to inform the user and the extension will not function until both parts are updated. This makes sure the project is properly updated and helps to prevent [some problems](https://github.com/filips123/PWAsForFirefox/issues/114#issuecomment-1002175674), but can be pretty confusing and annoying. To solve this, as said in the linked issue, I will change the update check to only require both parts to be updated on a new major SemVer version, which will happen every time a change in the communication protocol between the extension and the connector could cause problems for outdated versions. To implement those changes, the next version will probably be 2.0.0, and the future versions will follow the proposed versioning scheme.

Of course, the best way to make setup easier would be to integrate this project into official Firefox, but it is very unlikely that this will happen any time soon (the project would have to be almost completely rewritten, I don't have enough knowledge and time to do that, and Mozilla currently seems to be uninterested in supporting PWAs on desktop). On the bright side, I think having this as an independent project also has some advantages, such as easier customization (different addons and settings in PWA profiles, different widget order in titlebar...) and better support for multiple profiles, that built-in solutions probably wouldn't be able to offer.

### Other Improvements

I think the project just as "a tool to install PWAs" is mostly finished. So, apart from fixing existing problems I outlined in this discussion and implementing additional quality of life and utility features requested by users, it seems reasonable to also try supporting some additional PWA features that are currently only supported in Chromium-based browsers. However, many of those features are not yet standardized and the best form of "official" specification is a blog post by Google or Microsoft explaining how great they are (yeah, I really like how PWA APIs are developed). Additionally, I cannot nor want to support some of Chromium-specific APIs (such as access to the local filesystem, Bluetooth, NFC, USB, etc.), because that probably requires forking and modifying Firefox source (which I don't want to do), and I don't think giving websites those privileges is a good idea. For other, easier to implement and less dangerous APIs (like support for a protocol (#83) and ~~file~~ (actually no, because this requires access to local filesystem) handlers, share targets, as well as some cosmetic features such as titlebar manipulation), I think it might be worth experimenting with them.

What do you think about this plan? Which features would you like to see first? Which PWA APIs would you like to get supported?

### Bug Fixes

Unfortunately, there are still some pretty inconvenient issues, such as new windows being disconnected (#79), all PWAs in the same profile being merged on Linux and macOS (#80, #81) and a new tab being opened after closing all windows on macOS (#42). I think I partially fixed the first issue (and the fix will be available in the next version), but for a complete fix and other issues, forking Firefox would probably be necessary, which I don't want to do (it would make maintenance a lot harder, require us to create a custom infrastructure for building and updating custom Firefox fork, make it impossible to use the project with alternative Firefox versions...). Some of those issues (such as merging windows on Linux) are pretty generic and could be [fixed in Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1747722), while others (such as issues on macOS) are quite PWA-specific, so I'm not sure if a fix in the official Firefox would be appropriate. I will appreciate any help and ideas to fix those problems.

Apart from those bigger problems, have you noticed any other smaller bugs or inconveniences in PWAsForFirefox? Or maybe some glitches when using the modified Firefox UI? Some problems that I and other users have noticed are not working offline support (#82) or the inability to store the last window size and use it when reopening PWA. Are there any other bugs you found? On which of them should I focus first?

## Conclusion

I will appreciate it if you take some time to read and discuss those things or even help by contributing PRs, to make PWAsForFirefox a better project. And I wish you a happy new year!

**Comment by black7375** [comment]
### Documantation

> It seems it is still not enough to make PWAsForFirefox easier for less experienced users.

install? use? I don't know in some way, but it can be difficult if users are not a developer because it's mostly text.

### Easier Setup

- Easier changing of installation paths: Ideally I would use Firefox installed on my system as runtime, but I know there are technical issues. https://github.com/filips123/PWAsForFirefox/discussions/67
- Less intrusive updates:
  - And the best part is that it has automatic updates. (Like Firefox)
  - Currently, the biggest problem is that if we click the button of the add-on, we will be taken to the update page, and the add-on pop-up cannot be used.
  - Update alarm page is good, but as discussed, protocol versions are positive.
 
### Other Improvements

- Chromium like API: I think it's okay under the assumption that the permission is adjustable and that there are API on/off options.
- New Window: The current PWA site should be running.
- Extension Design: I think it's good to make the UI similar to Firefox.

### Bug Fixes

Personally, I prefer to do (#80, #81) first, but I don't have a strong opinion.
It is also good to start with the easy ones.

**Reply by githubDiversity** [reply]
first of all.
Please forgive my ignorance if I am not up to date yet.

WOW! thank you for your efforts. I am dismayed by firefox dropping/having dropped PWA support.

I have been warned already that Mozilla might become less reliable as they are moving towards google {trustworthy quote needed}.

I wish the world did not revolve around money. Only the sun.

Is there anything specific I can help with?

**Comment by githubDiversity** [comment]
Please do not get me wrong. I am an inspiring PWA developer focused on android chrome and IOS safari only. That is not to say I do not care about what happens to firefox.

We need a counter force otherwise google will dominate the landscape of which features are implemented or not in regards to the official Web API specs

**Reply by githubDiversity** [reply]
that's weird, I seem to be able to upvote and like my own comment

**Comment by githubDiversity** [comment]
All in all I might think we might yet again loose one field to a single player. This time browser technology related.

Google has even announced dropping http2/server push. My god!!!!!

**Comment by BambuSergio** [comment]
Backup and restore functions for save and restore all apps created.

Thank you. Great extension.

**Reply by filips123** [reply]
There isn't any easy function for this yet, but you can manually backup and restore all your apps. See [this](https://github.com/filips123/PWAsForFirefox/issues/130#issuecomment-1031558271) for more details.

**Comment by BambuSergio** [comment]
It's very easy and it's enough for me. Thank you very much for the answer!

Greetings.

**Comment by MenacingPerson** [comment]
Sync site data automatically (preferably both ways) like logins, cookies etc, because users might have been using that website for a long time before too.

Copy Extensions/Extension Data over

Copy over the settings in the Settings page for firefox

**Reply by MenacingPerson** [reply]
And, in the "Install this Web App", the default name should not be grey, but autofilled for better clarity on bad screens

**Reply by MenacingPerson** [reply]
![Before](https://user-images.githubusercontent.com/59611020/198870987-e96509b8-c35f-42f0-b4d6-0cf1945126b1.png)

![After](https://user-images.githubusercontent.com/59611020/198871031-e0fcd524-8687-473c-8f82-fd59ad476b1a.png)

**Reply by jpxsat** [reply]
This is exactly what I would expect... for the PWA to use the already-existent profile in firefox and not having to login again...

**Comment by KSXGitHub** [comment]
Open a PWA from right-click menu in the main browser. Open a PWA by clicking on a link in a PWA of the same profile.

**Comment by jpxsat** [comment]
Hi.

1. When installing the extension, I had to go through a lot of steps, and it seemed confusing/a lot even for me (I would call myself an experienced user that likes to help newbies) -- for wide adoption of the extension I see this being a major blocker for many people. If this could be improved, it would be marvelous. :)

2. When installing a PWA, I would expect for it to use the data that is already there on my browser. Case being, if I'm already logged in any website which I want to use as a PWA, when I install the PWA it should already “be there”, and not to start from a blank profile…

3. Reinforcing point 2, any new PWA should have access to all the extensions running in the current user profile, but the way is handled right now, by starting from scratch, any extension has to be added again… (ad blockers, password managers, so on so on)

For point 2, when choosing profiles it should be something like:
— “Current user profile” (the profile that is already running in Firefox)
— Create a new profile
— “Profile 1 you already created”
— Automatically create a new profile (I don't really understand this option, but it's already there)

**Reply by filips123** [reply]
> When installing the extension, I had to go through a lot of steps, and it seemed confusing/a lot even for me (I would call myself an experienced user that likes to help newbies) -- for wide adoption of the extension I see this being a major blocker for many people. If this could be improved, it would be marvelous. :)

What exactly would you like to be improved? I don't really know how could I made it easier, as it's already basically a step-by-step guide...

> [2] and [3]

Well, unfortunately it's not really possible to do this. Profiles for PWAs need to be separate because they include an additional code, change the behaviour of the browser, and provide additional settings/configuration that is different from the normal Firefox.

Maybe it could be possible to symlink just specific files from the main profile that contain site data, but I don't know how reliable that would be. I may try this in the future, but I'm not sure.

However, you can already create a profile template and use it to pre-configure settings for PWA profiles. It's still a one-time sync, but at least you don't have to configure everything multiple times if you use multiple PWA profiles.

> Automatically create a new profile (I don't really understand this option, but it's already there)

It automatically creates a new profiles with the name of the PWA you're installing. (While "Create a new profile" opens a popup where you need to manually enter the name and create it.)

**Reply by jpxsat** [reply]
Hi filips123 :)

Probably the expectation would be to “just” install a Firefox extension, and not going on a system-based extra program. I agree it's not hard, probably just more than what the average user expects.

From your answer, I think the profiles part is related to the same limitations as the above : you're building something on top of Firefox, so I can only imagine this is not easily done. Apologies if I sounded bitter, but rest assured that I'm loving this extension and I appreciate a lot this gorgeous piece of work and my comments are only here trying to help :)

Perhaps having to start a new session could be ok, but the part that felt really weird was to re-install extensions and going through Firefox inner settings again...

**Reply by filips123** [reply]
> Probably the expectation would be to “just” install a Firefox extension, and not going on a system-based extra program. I agree it's not hard, probably just more than what the average user expects.

Unfortunately, webextensions are very limited in what they can do, so implementing something like PWAs without an additional system program is impossible.

> Perhaps having to start a new session could be ok, but the part that felt really weird was to re-install extensions and going through Firefox inner settings again...

I know, but I don't think there is any good way to implement it... Maybe it is possible to do something with symlinks, but there are some potential problems:

* It may not be reliable in case Firefox changes how things are stored in the profile. It's not such a big problem as it's unlikely to happen, but it's still one thing to consider.

* Firefox normally doesn't allow two separate instances accessing the same profile and uses lock files to prevent that. We can get around this by symlinking just data files, without locks. The problem is that in this case, data corruption might happen, which would be quite bad. The good thing is that at least some data files use SQLite WAL mode, which (I think) prevents data corruption even with concurrent writes, so it may be possible to symlink at least these files. However, I would still need to check if this is the case and if everything works fine ([maybe related](https://sqlite.org/forum/info/66d7d9ae9a30f05b8cc1082726ac91899e9e1089384572ab5b79d16fae2a1f16)).

* For website data, cookies, session, etc., sharing data might actually be possible, as long as we make sure data corruption cannot happen.

* For settings, there are actually more problems. The first one is that most settings are just stored in a single file which probably doesn't have any protections against concurrent writes. The other problem is that there are settings that you will want to have differently configured in the main profile and in the PWA profiles (for example, toolbars layout, various things disabled in PWAs, additional settings for PWAs). So, even with protections for accessing the same file from multiple profiles, just symlinking won't work, because some settings would need to be different.

I might experiment with symlinks in the future, at least for SQLite files, I can't promise if anything useful will actually be done... If you have any other ideas how this could be implemented, please also share them.

**Comment by AlonsoK28** [comment]
Why does Firefox Android not allow the installation of PWAs, as chrome does?

**Reply by filips123** [reply]
It does... Just open the menu and click Install or Add to home screen.

**Reply by AlonsoK28** [reply]
In the case of the following PWA https://ng-price-checker-git-provider-autozone-calonso011.vercel.app

Firefox dont provide prompt to install PWA (Android 13)

![Screenshot_20240512-101720_Pixel Launcher](https://github.com/filips123/PWAsForFirefox/assets/37458700/2d409766-574d-42fe-951f-13aed029dacd)
![Screenshot_20240512-101900_Pixel Launcher](https://github.com/filips123/PWAsForFirefox/assets/37458700/3fb00073-12eb-4f02-9de4-f475a809acff)

According to this documentation Firefox need to install a extension to manage PWA on Android?  https://addons.mozilla.org/en-US/firefox/addon/pwas-for-firefox/


![Screenshot_20240512-002345_Pixel Launcher](https://github.com/filips123/PWAsForFirefox/assets/37458700/e59a87e7-8307-489c-84fd-c677a0a6f198)

**Reply by filips123** [reply]
This button should install PWA (or at least add it to home screen if it doesn't have manifest):

![slika](https://github.com/filips123/PWAsForFirefox/assets/16626308/273d248a-d148-4b97-9e01-83e2f9e1f2f6)

> According to this documentation Firefox need to install a extension to manage PWA on Android?

This extension is necessary only on desktop Firefox. Firefox for Android has PWAs built in.

**Reply by AlonsoK28** [reply]
This is the proof that PWA cannot be installed in Firefox

https://github.com/filips123/PWAsForFirefox/assets/37458700/156a56fe-c8d1-469c-aa2a-b4800635f91b

**Reply by filips123** [reply]
Web app manifest for https://ng-price-checker-git-provider-autozone-calonso011.vercel.app/ seems to have a few issues:

* It specifies `"./"` as `scope` and `start_url`, which will resolve relative to the manifest URL (probably `https://ng-price-checker-git-provider-autozone-calonso011.vercel.app/assets/manifest-files/` instead of `https://ng-price-checker-git-provider-autozone-calonso011.vercel.app/`). So, when the user opens the actual root URL, it will get detected as out of scope. Instead, `scope` and `start_url` should be set to `/`.

* Colors in `theme_color` and `background_color` should be prefixed with `#`.

Because of these issues, Firefox probably fails to parse the manifest and instead just adds the website as a normal shortcut.

**Comment by thivyan-studios** [comment]
Can we please sync extensions from the main browser to these PWA apps. The picture with Tree Style Tabs is the main browser, but, in a PWA such as YouTube, it does not install- instead I have to do it manually. This is just an example add-on; this occurs with any given add-on really.

Thanks for reading my request. Well done on your work thus far! Here's a coffee... ☕

![image](https://github.com/user-attachments/assets/50e4e50b-05c5-4769-b4f0-09a3ff8a5ae7)
![image1](https://github.com/user-attachments/assets/6948a447-112c-4d28-b4fc-0c3e68c0053a)

**Reply by filips123** [reply]
It might work if you sign into Firefox Account and sync the extensions. You could also use [a profile template](https://pwasforfirefox.filips.si/resources/profile-properties/#template) if you want to set up the extensions and settings once and use them for future all web app profiles.

Syncing extensions (and other settings and data) continuously with the main profile is quite hard to do and might not be completely reliable. I will check if I can do something about it in the future, but I'm not sure if it will work.

**Reply by thivyan-studios** [reply]
My bad. I didn't realize, but the extensions _do_ sync... they just take really long, lol. Perhaps there is something slowing down the process? If this suggestion doesn't make any sense, just remember it's coming from a beginner dev 😅

Look forward to seeing what you have in store for this add-on. 

bellic

**Reply by filips123** [reply]
Maybe it will be faster if you trigger it manually from the settings (`about:preferences#sync`) in the app browser.

**Reply by thivyan-studios** [reply]
Turns out I am also new to Firefox. Let me see if this helps...

**Comment by sleeptightAnsiC** [comment]
From my perspective as a user, the biggest problem is the lack of proper Tray support: https://github.com/filips123/PWAsForFirefox/issues/146

I'm trying to move as much to PWAs as possible, because the experience with "_native_" CEF/Electron apps is horrible. This is especially problematic with Discord which does not support Wayland, and running it under Xorg/Xwayland has its own issues. PWA is a perfect solution to most of those, but the fact that I don't get Tray icon/menu anymore, messes with my workflow.
